File: client/js/map.js
define(['jquery', 'area'], function ($, Area) {

  var Map = Class.extend({
    init: function (loadMultiTilesheets, game) {
      this.game = game;
      this.data = [];
      this.isLoaded = false;
      this.tilesetsLoaded = false;
      this.mapLoaded = false;
      this.loadMultiTilesheets = loadMultiTilesheets;

      var useWorker = !(this.game.renderer.mobile || this.game.renderer.tablet);

      this._loadMap(useWorker);
      this._initTilesets();
    },

    _checkReady: function () {
      if (this.tilesetsLoaded && this.mapLoaded) {
        this.isLoaded = true;
        if (this.ready_func) {
          this.ready_func();
        }
      }
    },

    _loadMap: function (useWorker) {
      var filepath = "maps/world_client.json";

      if (useWorker) {
        log.info("Loading map with web worker.");
        var worker = new Worker('js/mapworker.js');
        worker.postMessage(1);

        worker.onmessage = function (event) {
          var map = event.data;
          this._initMap(map);
          this.grid = map.grid;
          this.plateauGrid = map.plateauGrid;
          this.mapLoaded = true;
          this._checkReady();
        }.bind(this);
      } else {
        log.info("Loading map via Ajax.");
        $.get(filepath, function (data) {
          this._initMap(data);
          this._generateCollisionGrid();
          this._generatePlateauGrid();
          this.mapLoaded = true;
          this._checkReady();
        }.bind(this), 'json');
      }
    },

    _initTilesets: function () {
      var tileset1, tileset2, tileset3;

      if (!this.loadMultiTilesheets) {
        this.tilesetCount = 1;
        tileset1 = this._loadTileset('img/1/tilesheet.png');
      } else {
        if (this.game.renderer.mobile || this.game.renderer.tablet) {
          this.tilesetCount = 1;
          tileset2 = this._loadTileset('img/2/tilesheet.png');
        } else {
          this.tilesetCount = 2;
          tileset2 = this._loadTileset('img/2/tilesheet.png');
          tileset3 = this._loadTileset('img/3/tilesheet.png');
        }
      }

      this.tilesets = [tileset1, tileset2, tileset3];
    },

    _initMap: function (map) {
      this.width = map.width;
      this.height = map.height;
      this.tilesize = map.tilesize;
      this.data = map.data;
      this.blocking = map.blocking || [];
      this.plateau = map.plateau || [];
      this.musicAreas = map.musicAreas || [];
      this.collisions = map.collisions;
      this.high = map.high;
      this.animated = map.animated;

      this.doors = this._getDoors(map);
      this.checkpoints = this._getCheckpoints(map);
    },

    _getDoors: function (map) {
      var doors = {};

      _.each(map.doors, function (door) {
        var o;

        switch (door.to) {
          case 'u':
            o = Types.Orientations.UP;
            break;
          case 'd':
            o = Types.Orientations.DOWN;
            break;
          case 'l':
            o = Types.Orientations.LEFT;
            break;
          case 'r':
            o = Types.Orientations.RIGHT;
            break;
          default:
            o = Types.Orientations.DOWN;
        }

        doors[this.GridPositionToTileIndex(door.x, door.y)] = {
          x: door.tx,
          y: door.ty,
          orientation: o,
          cameraX: door.tcx,
          cameraY: door.tcy,
          portal: door.p === 1,
        };
      }.bind(this));

      return doors;
    },

    _loadTileset: function (filepath) {
      var tileset = new Image();

      tileset.src = filepath;

      log.info("Loading tileset: " + filepath);

      tileset.onload = function () {
        if (tileset.width % this.tilesize > 0) {
          throw Error("Tileset size should be a multiple of " + this.tilesize);
        }
        log.info("Map tileset loaded.");

        this.tilesetCount -= 1;
        if (this.tilesetCount === 0) {
          log.debug("All map tilesets loaded.")

          this.tilesetsLoaded = true;
          this._checkReady();
        }
      }.bind(this);

      return tileset;
    },

    ready: function (f) {
      this.ready_func = f;
    },

    tileIndexToGridPosition: function (tileNum) {
      var x = 0,
        y = 0;

      var getX = function (num, w) {
        if (num == 0) {
          return 0;
        }
        return (num % w == 0) ? w - 1 : (num % w) - 1;
      }

      tileNum -= 1;
      x = getX(tileNum + 1, this.width);
      y = Math.floor(tileNum / this.width);

      return {
        x: x,
        y: y
      };
    },

    GridPositionToTileIndex: function (x, y) {
      return (y * this.width) + x + 1;
    },

    isColliding: function (x, y) {
      if (this.isOutOfBounds(x, y) || !this.grid) {
        return false;
      }
      return (this.grid[y][x] === 1);
    },

    isPlateau: function (x, y) {
      if (this.isOutOfBounds(x, y) || !this.plateauGrid) {
        return false;
      }
      return (this.plateauGrid[y][x] === 1);
    },

    _generateCollisionGrid: function () {
      var tileIndex = 0;

      this.grid = [];
      for (var j, i = 0; i < this.height; i++) {
        this.grid[i] = [];
        for (j = 0; j < this.width; j++) {
          this.grid[i][j] = 0;
        }
      }

      _.each(this.collisions, function (tileIndex) {
        var pos = this.tileIndexToGridPosition(tileIndex + 1);
        this.grid[pos.y][pos.x] = 1;
      }.bind(this));

      _.each(this.blocking, function (tileIndex) {
        var pos = this.tileIndexToGridPosition(tileIndex + 1);
        if (this.grid[pos.y] !== undefined) {
          this.grid[pos.y][pos.x] = 1;
        }
      }.bind(this));

      log.info("Collision grid generated.");
    },

    _generatePlateauGrid: function () {
      var tileIndex = 0;

      this.plateauGrid = [];
      for (var j, i = 0; i < this.height; i++) {
        this.plateauGrid[i] = [];
        for (j = 0; j < this.width; j++) {
          if (_.include(this.plateau, tileIndex)) {
            this.plateauGrid[i][j] = 1;
          } else {
            this.plateauGrid[i][j] = 0;
          }
          tileIndex += 1;
        }
      }
      log.info("Plateau grid generated.");
    },

    /**
     * Returns true if the given position is located within the dimensions of the map.
     *
     * @returns {Boolean} Whether the position is out of bounds.
     */
    isOutOfBounds: function (x, y) {
      return isInt(x) && isInt(y) && (x < 0 || x >= this.width || y < 0 || y >= this.height);
    },

    /**
     * Returns true if the given tile id is "high", i.e. above all entities.
     * Used by the renderer to know which tiles to draw after all the entities
     * have been drawn.
     *
     * @param {Number} id The tile id in the tileset
     * @see Renderer.drawHighTiles
     */
    isHighTile: function (id) {
      return _.indexOf(this.high, id + 1) >= 0;
    },

    /**
     * Returns true if the tile is animated. Used by the renderer.
     * @param {Number} id The tile id in the tileset
     */
    isAnimatedTile: function (id) {
      return id + 1 in this.animated;
    },

    /**
     *
     */
    getTileAnimationLength: function (id) {
      return this.animated[id + 1].l;
    },

    /**
     *
     */
    getTileAnimationDelay: function (id) {
      var animProperties = this.animated[id + 1];
      if (animProperties.d) {
        return animProperties.d;
      } else {
        return 100;
      }
    },

    isDoor: function (x, y) {
      return this.doors[this.GridPositionToTileIndex(x, y)] !== undefined;
    },

    getDoorDestination: function (x, y) {
      return this.doors[this.GridPositionToTileIndex(x, y)];
    },

    _getCheckpoints: function (map) {
      var checkpoints = [];
      _.each(map.checkpoints, function (cp) {
        var area = new Area(cp.x, cp.y, cp.w, cp.h);
        area.id = cp.id;
        checkpoints.push(area);
      });
      return checkpoints;
    },

    getCurrentCheckpoint: function (entity) {
      return _.detect(this.checkpoints, function (checkpoint) {
        return checkpoint.contains(entity);
      });
    }
  });

  return Map;
});

File: client/dart/map.dart
library worldmap;

import "dart:convert";
import "dart:html";

import "base.dart";
import "checkpoint.dart";
import "door.dart";
import "entity.dart";
import "game.dart";
import "lib/gametypes.dart";
import 'position.dart';

class WorldMap extends Base {

  bool tilesetsLoaded = false;
  bool mapLoaded = false;
  bool loadMultiTilesheets;
  var map;
  List<List<int>> grid;
  List<List<int>> plateauGrid;
  Map<int, Door> doors = {};
  List<Checkpoint> checkpoints = [];
  List<ImageElement> tilesets = [];
  int tilesetCount;

  WorldMap(bool this.loadMultiTilesheets) {
    bool useWorker = !(Game.renderer.mobile || Game.renderer.tablet);
    this._loadMap(useWorker);
  }

  int get width => this.map["width"];
  int get height => this.map["height"];
  int get tilesize => this.map["tilesize"];
  get data => this.map["data"];
  get blocking => this.map.containsKey("blocking") ? this.map["blocking"] : [];
  get plateau => this.map.containsKey("plateau") ? this.map["plateau"] : [];
  get musicAreas => this.map.containsKey("musicAreas") ? this.map["musicAreas"] : [];
  get collisions => this.map["collisions"];
  get high => this.map["high"];
  get animated => this.map["animated"];

  bool get isLoaded => (this.tilesetsLoaded && this.mapLoaded);

  void _loadMap(bool useWorker) {
    String filepath = "maps/world_client.json";

    if (useWorker) {
      window.console.info("Loading map with web worker.");
      var worker = new Worker("js/mapworker.js");
      worker.postMessage(1);
      worker.onMessage.listen((event) {
        var map = event.data;
        this._initMap(map);
        this.grid = map['grid'];
        this.plateauGrid = map['plateauGrid'];
        this.mapLoaded = true;
        if (this.isLoaded) {
          this.trigger("Ready");
        }
        this._initTilesets();
      });
    } else {
      window.console.info("Loading map via Ajax.");
      HttpRequest.getString(filepath).then((String response) {
        var data = JSON.decode(response);
        this._initMap(data);
        this._generateCollisionGrid();
        this._generatePlateauGrid();
        this.mapLoaded = true;
        if (this.isLoaded) {
          this.trigger("Ready");
        }
        this._initTilesets();
      }).catchError((Error error) {
        window.console.error("Failed loading map via AJAX. Error: ${error}");
      });
    }
  }

  void _initMap(map) {
    this.map = map;

    this.map['doors'].forEach((doorData) {
      Orientation o;

      switch (doorData["to"]) {
        case 'u':
          o = Orientation.UP;
          break;
        case 'd':
          o = Orientation.DOWN;
          break;
        case 'l':
          o = Orientation.LEFT;
          break;
        case 'r':
          o = Orientation.RIGHT;
          break;
        default:
          o = Orientation.DOWN;
      }

      Door door = new Door(new Position(doorData['tx'], doorData['ty']), o, new Position(doorData['tcx'], doorData['tcy']), doorData['p'] == 1);
      doors[this.gridPositionToTileIndex(door.position)] = door;
    });

    this.map['checkpoints'].forEach((cp) {
      Checkpoint checkpoint = new Checkpoint(cp['id'], cp['x'], cp['y'], cp['w'], cp['h']);
      this.checkpoints.add(checkpoint);
    });
  }

  void _initTilesets() {
    ImageElement tileset1, tileset2, tileset3;

    if (!this.loadMultiTilesheets) {
      this.tilesetCount = 1;
      tileset1 = this._loadTileset('img/1/tilesheet.png');
    } else {
      if (Game.renderer.mobile || Game.renderer.tablet) {
        this.tilesetCount = 1;
        tileset2 = this._loadTileset('img/2/tilesheet.png');
      } else {
        this.tilesetCount = 2;
        tileset2 = this._loadTileset('img/2/tilesheet.png');
        tileset3 = this._loadTileset('img/3/tilesheet.png');
      }
    }

    this.tilesets = [tileset1, tileset2, tileset3];
  }

  ImageElement _loadTileset(String filepath) {
    ImageElement tileset = new ImageElement();
    tileset.src = filepath;

    window.console.info("Loading tileset: ${filepath}");

    tileset.onLoad.listen((e) {
      if (tileset.width % this.tilesize > 0) {
        throw "Tileset size should be a multiple of ${this.tilesize}";
      }
      window.console.info("Map tileset loaded.");

      this.tilesetCount -= 1;
      if (this.tilesetCount == 0) {
        window.console.info("All map tilesets loaded.");

        this.tilesetsLoaded = true;
        if (this.isLoaded) {
          this.trigger("Ready");
        }
      }
    });

    return tileset;
  }

  Position tileIndexToGridPosition(int tileNum) {
    int x = 0;
    int y = 0;

    tileNum--;

    if (x != 0) {
      if (((tileNum + 1) % this.width) == 0) {
        x = this.width - 1;
      } else {
        x = ((tileNum + 1) % this.width) - 1;
      }
    }

    y = (tileNum / this.width).floor();

    return new Position(x, y);
  }

  int gridPositionToTileIndex(Position position) => (position.y * this.width) + position.x + 1;

  bool isColliding(Position pos) =>
    !this.isOutOfBounds(pos)
    && this.grid != null
    && this.grid[pos.y][pos.x] == 1;

  bool isPlateau(Position pos) =>
    !this.isOutOfBounds(pos)
    && this.plateauGrid != null
    && this.plateauGrid[pos.y][pos.x] == 1;

  void _generateCollisionGrid() {
    var tileIndex = 0;

    this.grid = [];
    for (var j, i = 0; i < this.height; i++) {
      this.grid[i] = []..fillRange(0, this.width, 0);
    }

    this.collisions.forEach((int tileIndex) {
      Position pos = this.tileIndexToGridPosition(tileIndex + 1);
      this.grid[pos.y][pos.x] = 1;
    });

    this.blocking.forEach((int tileIndex) {
      Position pos = this.tileIndexToGridPosition(tileIndex + 1);
      this.grid[pos.y][pos.x] = 1;
    });

    window.console.info("Collision grid generated.");
  }

  void _generatePlateauGrid() {
    var tileIndex = 0;

    this.plateauGrid = [];
    for (var j, i = 0; i < this.height; i++) {
      this.plateauGrid[i] = [];
      for (j = 0; j < this.width; j++) {
        this.plateauGrid[i][j] = this.plateau.contains(tileIndex) ? 1 : 0;
        tileIndex++;
      }
    }

    window.console.info("Plateau grid generated.");
  }

  bool isOutOfBounds(Position pos) =>
    (pos.x < 0 || pos.x >= this.width || pos.y < 0 || pos.y >= this.height);

  /**
   * Returns true if the given tile id is "high", i.e. above all entities.
   * Used by the renderer to know which tiles to draw after all the entities
   * have been drawn.
   *
   * @param {Number} id The tile id in the tileset
   * @see Renderer.drawHighTiles
   */
  bool isHighTile(int id) => this.high.contains(id + 1);

  /**
   * Returns true if the tile is animated. Used by the renderer.
   * @param {Number} id The tile id in the tileset
   */
  bool isAnimatedTile(int id) => this.animated.containsKey(id + 1);

  int getTileAnimationLength(int id) => this.animated[id + 1]["l"];

  int getTileAnimationDelay(id) {
    if (this.animated[id + 1].containsKey("d")) {
      return this.animated[id + 1]["d"];
    }

    return 100;
  }

  bool isDoor(Position position) =>
    this.doors.containsKey(this.gridPositionToTileIndex(position));

  Door getDoorDestination(Position position) => this.doors[this.gridPositionToTileIndex(position)];

  Checkpoint getCurrentCheckpoint(Entity entity) {
    for (final checkpoint in this.checkpoints) {
      if (checkpoint.contains(entity)) {
        return checkpoint;
      }
    }

    return null;
  }
}

